doc/plan-for-new-modules-implementation.md

```diff
@@ -0,0 +1,100 @@
+
+* `createRequireFunction` ([nodejs/node#19360](https://github.com/nodejs/node/pull/19360)) is the only way to import CommonJS into an ES module, for now.
+  - `import.meta.require` does not fail early enough.
+  - Hold off on `import` statements for CommonJS until more progress is made on the dynamic modules spec.
+
+* `import` statements will only support files with an `.mjs` extension, and will import only ES modules, for now.
+  - In a later phase, the intention is to move forward with format databases to map extensions and support multiple use cases.
+  - No JSON or native modules; `createRequireFunction` can be used to get these.
+
+
+### How will we get from where we are to Phase 1
+
+* Remove support in the `import` statement of formats other than ESM:
+  - No CommonJS.
+  - No JSON.
+  - No native modules.
+  - Implemented in https://github.com/nodejs/ecmascript-modules/pull/3
+
+* Remove dynamic path searching:
+  - No extension adding.
+  - No directory resolution, including no support for `index.js` or `index.mjs`.
+  - Still support the `main` field, but it must specify a full filename.
+  - Implemented in https://github.com/nodejs/ecmascript-modules/pull/2
+
+* Add `createRequireFunction`.
+  - Implemented in https://github.com/nodejs/node/pull/19360
+
+
+## Features Expected in Later Phases
+
+These features are agreed upon, but make sense to include in a later phase of development; or there is agreement on the need for this feature, but a lack of consensus on the implementation details:
+
+* Browser-compatible specifier resolution ([#109](https://github.com/nodejs/modules/issues/109)), a.k.a. bare imports.
+  - Implementation to be discussed, possibly based on [package name maps](https://github.com/domenic/package-name-maps).
+
+* `import.meta.url`.
+
+* Dynamic `import()`.
+
+* User-configurable map for file extensions to parse goals, a.k.a. `mimes` field ([#160](https://github.com/nodejs/modules/pull/160)).
+  - This can expand the uses of the `import` statement to import ES modules from `.js` files, as well as support JSON and other formats.
+  - Common formats like JSON or WASM will have default mappings.
+  - Need to work out how this functionality interacts with loaders.
+
+* Dynamic path searching.
+  - Requiring the full path is an issue for tooling and a long term solution is required.
+  - Migration strategies also have issue with this.
+  - Need to resolve issues related to building this in a browser-compatible way.
+
+* Multi-mode packages ([#94](https://github.com/nodejs/modules/issues/94)).
+
+* Loaders ([#82](https://github.com/nodejs/modules/issues/82)), ([#96](https://github.com/nodejs/modules/issues/96)).
+  - Implementation to be determined.
+
+* ESM in executable files ([#152](https://github.com/nodejs/modules/issues/152)).
+
+* Callable resolver ([#157](https://github.com/nodejs/modules/issues/157)).
+
+* Mock modules (injection) ([#98](https://github.com/nodejs/modules/issues/98)).
+
+
+## Other Features
```

Comment:
This seems odd and ripe for controversy tbh

I'd like us to reach consensus on a minimal implementation as soon as possible so we can begin iterating over specific combinations of features. I don't think that this list of "expected features" vs "other features" is based on any specific consensus and is likely to distract from getting consensus on this document.

I personally think we are better off focusing on "what the kernel is" and "what specific changes get us there". What's next should be a planning phase next, or we simply won't get anything done.

Reply:
On the one hand I see your point, and feel like sure, we can just cut the section. But on the other, I think it’s useful to list out features that are blocked with reasons why they’re blocked, like all the ones waiting on spec updates. This could be a separate or later document though. What do you think?